### ISSUE TITLE
On will/didFocus subscribe, stop firing the listener if current screen is focused

### ISSUE THREAD

**Issue by slorber** [react-navigation/rfcs#72]
Hi,

Today, when we add a will/didFocus listener on a screen, if that screen is currently focused, the listener will fire.

I think that behavior was implemented at a time where `navigation.isFocused()` didn't exist yet and does not make sense anymore and is complicating the library usage.

For example, it makes it impossible to unsubscribe/resubscribe the same listener (or eventually closure/arrow function) without any side effect, which complicates the implementation of consuming code which need to ensure to register "stable listeners"
- https://github.com/react-navigation/react-navigation-core/pull/17/files#diff-ae2970cfcec67f1f42078a839a0a1f3f
- https://github.com/react-navigation/react-navigation-hooks/pull/8/files
- https://github.com/react-navigation/react-navigation-hooks/pull/8#issuecomment-442055111
- https://github.com/react-navigation/react-navigation/issues/5058

Also, one major usecase for navigation events is to refresh the screen data on focus.
The problem is that we use data loading systems that may already perform a fetch on mount (react-apollo, react-refetch...), so wiring `onWillFocus={() => refetch()}` will actually trigger a duplicate fetch on mount.
As mentionned during the implementation of `<NavigationEvents />` [here](https://github.com/react-navigation/rfcs/issues/41#issuecomment-387007440), I solved this problem on my app with this kind of code:

```js
handleWillFocus = (() => {
    let count = 0;
    return payload => {
            if ( payload.action.type === NavigationActions.BACK ) {
              return; // We don't want to refresh on back so that user does not loose pagination state
            }
            if ( count > 0 ) {
              this.refreshTabData()
            }
           count++;
    }
  })();
```

Removing the `willFocus` event after subscribe of a mounted screen will solve that problem for all people that are trying to refetch on screen focus. If the behavior is intended it can still added back in componentDidMount anyway.

-----------------------

This is a breaking change and users should perform these migrations:

1) If user wants to fire something on mount if the screen is focused, he can implement that with `navigation.isFocused()` in `componentDidMount`

2) If user wants to track focused state, he can use `withNavigationFocus` (also implemented later) or init his state with `navigation.isFocused()`

-----------------------

If this is validated I'd be happy to send a PR to all the concerned projects of the system (core, NavigationEvents, hooks)

**Comment by brentvatne** [comment]
> Today, when we add a will/didFocus listener on a screen, if that screen is currently focused, the listener will fire.


can you provide a snack that demonstrates clearly what is happening here?

**Comment by slorber** [comment]
Hi @brentvatne 

# TLDR

I was not totally correct in my assumptions above. 


**Infinite loop**.

When using declarative event comp or hooks, it's simpler to unsub/resub to ensure the latest closure is run. But doing so on component updates does something equivalent of the following:

```js
  componentDidMount() {
    this.props.navigation.addListener('willFocus', () => {
      this.incrementCounter();
      this.props.navigation.addListener('willFocus', () => {
        this.incrementCounter();
      });
    });
  }
```
The infinite loop problem is because when adding a 2nd `willFocus` listener during the 1st `willFocus` will actually trigger both listeners (counter=2) while I think it should only execute the 1st

This problem was also present in early versions of Redux and I can look this to solve this independently and ensure that only the 1st listener is called. The 2nd listener can be called only the screen focuses again. Changing this behavior only is a smaller breaking change and will simplify the implementation of hooks and NavigationEvents

**Do we really want to have this listener fired on mount?**

For me the main usecase to `willFocus` was mostly to track focused screen before, but we have better alternatives now (withNavigationFocus, navigation.isFocused()...). The main usecase that remains for me is to refresh api data on screen focus. But generally the tools we use already do the fetching on mount (like Apollo) so having this event firing will trigger a "double fetch" on mount. Do we really want this event to fire when the initial navigator screen is mounting? 

I don't know how users are using this listener, maybe my assumptions are incorrect, and it would be  a larger breaking change.



# Details

Here is a snack that shows the infinite loop problem the current solution can actually trigger: https://snack.expo.io/@slorber/focus-event

The original issue reporting this is https://github.com/react-navigation/react-navigation/issues/5058


This loop happens because:
- The listener triggers a state change
- We unsub/resub on every update
- The resub triggers the listener

My snack may feel convoluted and the unsub/resub unnecessary, but when using declarative approachs like <NavigationEvents/> or hooks, particularly when providing closures as listeners, it's important to ensure the last closure provided is run because a closure capture external variables at creating time, and those variables may change over time. Unsub/resub permits to easily ensure the last closure is run.


Actually I think my initial description of the problem is not totally correct. It seems unsub/resub happening inside a "didFocus" is actually the problem. The button in the snack does show that when the screen is already focused, and we add a listener, this listener will actually not be fired even if the screen is focused (where I was wrong).

As far as I remember, Redux had this issue too and solved it by creating a listener array copy before iterating on them, so that on resub, the newly added listener won't be executed before next focus. Doing something similar can make sense here to ensure that we don't encounter any infinite loop.

The following should only print 1 imho, currently it prints 2: https://snack.expo.io/@slorber/focus-event-2
This would solve the infinite loop issue, but would not solve the issue where the user might not necessarily want this listener to fire on mount (because it may be wired to a "refresh api data" method)
